resource/file/file_test.go

```diff
@@ -0,0 +1,254 @@
+		{File{Destination: filepath.Join(tmpDir, "file-content"), Mode: mode(0666), Content: []byte("converge/content test\n"), State: "present"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "dir-A"), Mode: mode(0700), State: "present", Type: "directory"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "dir-missing"), State: "absent"}, nil, false},
+		{File{Destination: filepath.Join(tmpDir, "dir-A/dir-B"), State: "present", Type: "directory"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "dir-C/dir-D/dir-E/dir-F"), State: "present", Type: "directory"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "file-symlink"), Target: filepath.Join(tmpDir, "file-content"), Type: "symlink"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "file-hardlink"), Target: filepath.Join(tmpDir, "file-test"), Type: "hardlink"}, nil, true},
+	}
+
+	var modify = []Case{
+		{File{Destination: filepath.Join(tmpDir, "file-test"), Mode: mode(0700), State: "present"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "file-content"), Mode: mode(0600), Content: []byte("converge/updated content\n"), State: "present"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "dir-A"), Mode: mode(0750), State: "present", Type: "directory"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "dir-C/dir-D/dir-E/dir-F"), Force: true, State: "present", Type: "file"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "file-symlink"), Force: true, Target: filepath.Join(tmpDir, "dir-C/dir-D/dir-E/dir-F"), Type: "symlink"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "file-hardlink"), Mode: mode(0550), Force: true, Target: filepath.Join(tmpDir, "file-content"), Type: "hardlink"}, nil, true},
+	}
+
+	var delete = []Case{
+		{File{Destination: filepath.Join(tmpDir, "file-test"), Mode: mode(0750), State: "absent"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "file-missing"), State: "absent"}, nil, false},
+		{File{Destination: filepath.Join(tmpDir, "file-content"), Content: []byte("converge/content test\n"), State: "absent"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "dir-A"), Mode: mode(0700), State: "absent", Type: "directory"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "dir-missing"), State: "absent"}, nil, false},
+		{File{Destination: filepath.Join(tmpDir, "dir-A/dir-B"), State: "absent", Type: "directory"}, nil, false},       //parent dir removed first
+		{File{Destination: filepath.Join(tmpDir, "dir-C"), State: "absent", Type: "directory", Force: true}, nil, true}, //remove child dirs
+		{File{Destination: filepath.Join(tmpDir, "file-symlink"), State: "absent", Target: filepath.Join(tmpDir, "file-content"), Type: "symlink"}, nil, true},
+		{File{Destination: filepath.Join(tmpDir, "file-hardlink"), State: "absent", Target: filepath.Join(tmpDir, "file-test"), Type: "hardlink"}, nil, true},
+	}
+
```

Comment:
It looks like these test cases check successes; we probably need some tests for the different ways failure can occur.